File: mobile/ios/PrestamosApp/Utils/OfflineHelper.swift
import Foundation

/**
 * Helper para manejo de datos offline en iOS
 */
class OfflineHelper {
    private let userDefaults = UserDefaults.standard
    private let keyPendingActions = "offline_pending_actions"
    private let keyLastSync = "offline_last_sync"
    
    struct PendingAction: Codable, Equatable {
        let tipo: String
        let datos: String
        let timestamp: TimeInterval
        var intentos: Int
    }
    
    /**
     * Guardar acción pendiente para sincronizar
     */
    func guardarAccionPendiente(tipo: String, datos: Codable) {
        var acciones = obtenerAccionesPendientes()
        
        let datosJson: String
        if let jsonData = try? JSONEncoder().encode(datos),
           let jsonString = String(data: jsonData, encoding: .utf8) {
            datosJson = jsonString
        } else {
            return
        }
        
        let accion = PendingAction(
            tipo: tipo,
            datos: datosJson,
            timestamp: Date().timeIntervalSince1970,
            intentos: 0
        )
        
        acciones.append(accion)
        guardarAcciones(acciones)
    }
    
    /**
     * Obtener acciones pendientes
     */
    func obtenerAccionesPendientes() -> [PendingAction] {
        guard let data = userDefaults.data(forKey: keyPendingActions),
              let acciones = try? JSONDecoder().decode([PendingAction].self, from: data) else {
            return []
        }
        return acciones
    }
    
    /**
     * Eliminar acción pendiente
     */
    func eliminarAccionPendiente(_ accion: PendingAction) {
        var acciones = obtenerAccionesPendientes()
        acciones.removeAll { $0 == accion }
        guardarAcciones(acciones)
    }
    
    /**
     * Incrementar intentos de una acción
     */
    func incrementarIntentos(_ accion: PendingAction) {
        var acciones = obtenerAccionesPendientes()
        if let index = acciones.firstIndex(of: accion) {
            acciones[index].intentos += 1
        }
        guardarAcciones(acciones)
    }
    
    /**
     * Guardar acciones
     */
    private func guardarAcciones(_ acciones: [PendingAction]) {
        if let data = try? JSONEncoder().encode(acciones) {
            userDefaults.set(data, forKey: keyPendingActions)
        }
    }
    
    /**
     * Guardar última sincronización
     */
    func guardarUltimaSincronizacion() {
        userDefaults.set(Date().timeIntervalSince1970, forKey: keyLastSync)
    }
    
    /**
     * Obtener última sincronización
     */
    func obtenerUltimaSincronizacion() -> TimeInterval {
        return userDefaults.double(forKey: keyLastSync)
    }
    
    /**
     * Verificar si hay datos pendientes
     */
    func hayDatosPendientes() -> Bool {
        return !obtenerAccionesPendientes().isEmpty
    }
    
    /**
     * Limpiar acciones antiguas (más de 7 días)
     */
    func limpiarAccionesAntiguas() {
        let sieteDiasAtras = Date().timeIntervalSince1970 - (7 * 24 * 60 * 60)
        let acciones = obtenerAccionesPendientes()
        let accionesValidas = acciones.filter { $0.timestamp > sieteDiasAtras }
        guardarAcciones(accionesValidas)
    }
}

File: mobile/android/app/src/main/java/com/erp/prestamos/utils/OfflineHelper.java
package com.erp.prestamos.utils;

import android.content.Context;
import android.content.SharedPreferences;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.List;

/**
 * Helper para manejo de datos offline
 */
public class OfflineHelper {
    private static final String PREFS_NAME = "offline_data";
    private static final String KEY_PENDING_ACTIONS = "pending_actions";
    private static final String KEY_LAST_SYNC = "last_sync";
    
    private SharedPreferences prefs;
    private Gson gson;
    
    public OfflineHelper(Context context) {
        prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        gson = new Gson();
    }
    
    /**
     * Guardar acción pendiente para sincronizar
     */
    public void guardarAccionPendiente(String tipo, Object datos) {
        List<PendingAction> acciones = obtenerAccionesPendientes();
        
        PendingAction accion = new PendingAction();
        accion.tipo = tipo;
        accion.datos = gson.toJson(datos);
        accion.timestamp = System.currentTimeMillis();
        accion.intentos = 0;
        
        acciones.add(accion);
        
        guardarAcciones(acciones);
    }
    
    /**
     * Obtener acciones pendientes
     */
    public List<PendingAction> obtenerAccionesPendientes() {
        String json = prefs.getString(KEY_PENDING_ACTIONS, "[]");
        Type listType = new TypeToken<List<PendingAction>>(){}.getType();
        return gson.fromJson(json, listType);
    }
    
    /**
     * Eliminar acción pendiente
     */
    public void eliminarAccionPendiente(PendingAction accion) {
        List<PendingAction> acciones = obtenerAccionesPendientes();
        acciones.remove(accion);
        guardarAcciones(acciones);
    }
    
    /**
     * Incrementar intentos de una acción
     */
    public void incrementarIntentos(PendingAction accion) {
        List<PendingAction> acciones = obtenerAccionesPendientes();
        for (PendingAction a : acciones) {
            if (a.equals(accion)) {
                a.intentos++;
                break;
            }
        }
        guardarAcciones(acciones);
    }
    
    /**
     * Guardar acciones
     */
    private void guardarAcciones(List<PendingAction> acciones) {
        String json = gson.toJson(acciones);
        prefs.edit().putString(KEY_PENDING_ACTIONS, json).apply();
    }
    
    /**
     * Guardar última sincronización
     */
    public void guardarUltimaSincronizacion() {
        prefs.edit().putLong(KEY_LAST_SYNC, System.currentTimeMillis()).apply();
    }
    
    /**
     * Obtener última sincronización
     */
    public long obtenerUltimaSincronizacion() {
        return prefs.getLong(KEY_LAST_SYNC, 0);
    }
    
    /**
     * Verificar si hay datos pendientes
     */
    public boolean hayDatosPendientes() {
        List<PendingAction> acciones = obtenerAccionesPendientes();
        return !acciones.isEmpty();
    }
    
    /**
     * Limpiar acciones antiguas (más de 7 días)
     */
    public void limpiarAccionesAntiguas() {
        List<PendingAction> acciones = obtenerAccionesPendientes();
        long sieteDiasAtras = System.currentTimeMillis() - (7 * 24 * 60 * 60 * 1000);
        
        List<PendingAction> accionesValidas = new ArrayList<>();
        for (PendingAction accion : acciones) {
            if (accion.timestamp > sieteDiasAtras) {
                accionesValidas.add(accion);
            }
        }
        
        guardarAcciones(accionesValidas);
    }
    
    /**
     * Clase para acciones pendientes
     */
    public static class PendingAction {
        public String tipo;
        public String datos;
        public long timestamp;
        public int intentos;
        
        @Override
        public boolean equals(Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            PendingAction that = (PendingAction) obj;
            return tipo.equals(that.tipo) && 
                   datos.equals(that.datos) && 
                   timestamp == that.timestamp;
        }
    }
}

